### PR TITLE
Slava br

### PR DESCRIPTION
**Pull request by MeshchaninovViacheslav**
I've conducted experiments with encoder model for text diffusion. The main focus was on learning padding encodings. The result is that the best solution is to zero all paddings. It gives the best score on RocStories, however, slightly worse than regular mode on QQP

### CODE REVIEW

File: autoencoder/autoencoder.py [hunk 1]
```diff
@@ -7,25 +7,27 @@
...
-            n_layer=4,
+            num_latents=config.decoder.num_latents,
+            embedding_dim=config.decoder.embedding_dim,
+            hidden_size=config.decoder.hidden_size,
+            n_heads=config.decoder.n_heads,
+            n_layers=config.decoder.n_layers,
+            dropout_p=config.decoder.dropout_p,
         )
-        self.projector = nn.Linear(768, 30522, bias=False)
+        self.projector = nn.Linear(config.projector.hidden_size, config.projector.vocab_size, bias=False)
```
Comment (M0RJIQUE) [inline autoencoder/autoencoder.py:30]: Why setting `bias=False`?

File: autoencoder/config.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    encoder = config.encoder = ml_collections.ConfigDict()
+    encoder.name = "bert-base-uncased"
+    encoder.config = AutoConfig.from_pretrained(encoder.name)
+    
+    compressor = config.compressor = ml_collections.ConfigDict()
+    compressor.num_latents = 32
+    compressor.embedding_dim = 768
+    compressor.hidden_size = 768
+    compressor.n_heads = 12
+    compressor.n_layers = 4
```
Comment (M0RJIQUE) [inline autoencoder/config.py:15]: LD4TG utilizes 3 layers for both autoencoding modules, I think for consistency purposes it is wise to follow their implementation

File: autoencoder/config.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    compressor.dropout_p = 0.1
+
+    decoder = config.decoder = ml_collections.ConfigDict()
+    decoder.num_latents = 64
+    decoder.embedding_dim = 768
+    decoder.hidden_size = 768
+    decoder.n_heads = 12
+    decoder.n_layers = 4
```
Comment (M0RJIQUE) [inline autoencoder/config.py:23]: Same as for the number of compressor layers

File: model/decoder.py [hunk 4]
```diff
@@ -1,29 +1,18 @@
...
-            self.fc = nn.Linear(config.hidden_size, config.vocab_size)
-            self.net = lambda x: self.fc(self.bert(x).last_hidden_state)
-
-        elif mode == 'mlm':
-            self.cls = BertOnlyMLMHead(config)
-            self.net = lambda x: self.cls(x)
-        else:
-            print('Unknown decoder mode', flush=True)
-            raise
+        bert_config.num_hidden_layers = 6
```
Comment (M0RJIQUE) [inline model/decoder.py:12]: I think it is better to write this constant in config for consistency purposes

File: model/train_decoder.py [hunk 5]
```diff
@@ -111,26 +103,23 @@ def train(config, encoder, decoder, tokenizer):
...
-        lr=5e-5,
+        lr=1e-4,
         weight_decay=0.001,
         betas=(0.9, 0.98),
     )
 
     eval_freq = 100
     step = 0
-    epochs = 10
+    epochs = 4
```
Comment (M0RJIQUE) [inline model/train_decoder.py:122]: Maybe we should put this into config too

Comment (MeshchaninovViacheslav) [reply model/train_decoder.py:122]: Agree
